Adding 2 improvements related to convective transport of trace gases into WRF-Chem

liyunyao opened the pull request:
TYPE: new feature

KEYWORDS: GF-Cumulus Scheme, Convective transport of trace gases, Wet scavenging, subgrid scale

SOURCE: Yunyao Li (University of Maryland)

DESCRIPTION OF CHANGES:
I add the trace gases and aerosol subgrid-scale convective transport (Grell and Freitas, 2014; Li et al., 2018), subgrid-scale wet scavenging (including the improvements of ice retention factors and the conversion ratio of cloud water to rainwater, Li et al., 2019), and aqueous chemistry to the GF cumulus parameterization.
I also put some chem codes to share/ directory. So some of the chemistry process can be called in the physics part, and the model doesn't need to be compiled twice to get everything work.

ISSUE: N/A

LIST OF MODIFIED FILES:
M chem/Makefile
M chem/chem_driver.F
M chem/depend.chem
D chem/module_HLaw.F
D chem/module_ctrans_aqchem.F
M chem/module_input_chem_data.F
M phys/Makefile
M phys/module_cu_gf_deep.F
M phys/module_cu_gf_sh.F
M phys/module_cu_gf_wrfdrv.F
M phys/module_cumulus_driver.F
M share/Makefile

Note: chem/module_HLaw.F and chem/module_ctrans_aqchem.F are moved to share/
Two new files are added: 
share/module_chem_share.F
phys/module_cu_gf_ctrans.F

TESTS CONDUCTED:
Li, Y., Pickering, K.E., Barth, M.C., Bela, M.M., Cummings, K.A., and Allen, D.J. (2019). Wet Scavenging in WRF-Chem Simulations of Parameterized Convection for a Severe Storm during the DC3 Field Campaign, J. Geophys. Res. Atmos. 124, 7413–7428. https://doi.org/10.1029/2019JD030484.

Li, Y., Pickering, K.E., Barth, M.C., Bela, M.M., Cummings, K.A., and Allen, D.J. (2018). Evaluation of Parameterized Convective Transport of Trace Gases in Simulation of Storms Observed During the DC3 Field Campaign. J. Geophys. Res. Atmos. 123, 11,238-11,261.

weiwangncar commented:
@liyunyao Your PR will trigger an automated test. After a couple of hours, you should receive an email from jenkins@no-reply.scalacomputing.com (check your Spam folder too). The message may be obvious to you to figure out if all of the tests are ok. If you have questions, post it here. If you're convinced that the test has passed, update the PR under 'Tests Conducted'.

davegill commented:
@liyunyao 
I am rebasing this this PR to release-v4.2 from the develop branch

davegill commented:
@liyunyao 
There are 14 files that are modified, only 12 are listed. Please update the PR message.

davegill commented:
@liyunyao 
> Note: chem/module_HLaw.F and chem/module_ctrans_aqchem.F are moved to share/

How does the build work with chem-only code in the share directory?

liyunyao commented:
> @liyunyao
> 
> > Note: chem/module_HLaw.F and chem/module_ctrans_aqchem.F are moved to share/
> 
> How does the build work with chem-only code in the share directory?

Hi Dave, I have fixed the bugs. I will re-submit the new version soon. I add "#if ( WRF_CHEM == 1 )" in the .F files I modified under /share/

weiwangncar commented:
@liyunyao You don't need to close this PR and redo another. You can just make changes to this PR from your local working directory.

ravanah commented:
This code is reviewed by Georg and other team members. Please wait for our comments.

jordanschnell commented:
@liyunyao  - This looks great to us. One thing we would appreciate would be to make c0 a namelist variable (though you may want to call it something more descriptive). @ravanah

davegill commented on chem/chem_driver.F:
```diff
@@ -1132,6 +1132,9 @@ REAL(KIND=8) :: real_time_r8                                 !ext. function in a
 !
 !
       if( config_flags%cu_physics>0 .and. config_flags%chem_conv_tr>0 &
+!++ GF CTRAN ++ lyy 01/2020
```
Remove these "!++ GF CTRAN ++ lyy" lines

davegill commented:
@liyunyao 
If you remove module_HLaw.F from the chem dir, it should be now in the share dir (based on your makefile mods). Why is that file not listed?

davegill commented:
@liyunyao 
Same question about aqueous_ctran

davegill commented on chem/module_input_chem_data.F:
```diff
@@ -52,7 +52,9 @@ MODULE module_input_chem_data
    USE module_get_file_names, ONLY : eligible_file_name, number_of_eligible_files, unix_ls
    USE module_model_constants
-
+!++ move get_last_gas to share lyy 01/2020
```
zap these comments that refer to where files are located

davegill commented on chem/module_input_chem_data.F:
```diff
@@ -2389,89 +2391,89 @@ SUBROUTINE cv_mmdd_jday ( YY, MM, DD, JDAY)
-     end select
-
-   END FUNCTION get_last_gas
+!++ move get_last_gas to share lyy 01/2020
```
If the source is removed, do not just comment it out - remove it

davegill commented on phys/Makefile:
```diff
@@ -65,6 +65,7 @@ MODULES = \
 	module_cu_tiedtke.o\
 	module_cu_ntiedtke.o\
 	module_cu_gd.o \
+        module_cu_gf_ctrans.o \
```
follow the indenting in the original file, whether it is spaces or tabs

davegill commented on phys/module_cu_gf_deep.F:
```diff
@@ -1,4 +1,9 @@
 MODULE module_cu_gf_deep
+!++ GF CTRAN ++ lyy 01/2020
```
remove these comment lines

davegill commented on phys/module_cu_gf_deep.F:
```diff
@@ -97,6 +102,15 @@ SUBROUTINE CUP_gf(        &
 #if ( WRF_DFI_RADAR == 1 )
               ,do_capsuppress,cap_suppress_j    &             
 #endif
+!++ GF CTRAN ++ lyy 01/2020
```
zap comment lines

davegill commented on phys/module_cu_gf_deep.F:
```diff
@@ -170,7 +184,24 @@ SUBROUTINE CUP_gf(        &
        real                                                            &
         ,intent (in   )                   ::                           &
         dtime
+!++ GF CTRAN ++ lyy 01/2020
```
remove comment lines

davegill commented on phys/module_cu_gf_deep.F:
```diff
@@ -3448,7 +3521,8 @@ SUBROUTINE cup_up_moisture(name,ierr,z_cup,qc,qrc,pw,pwav,     &
             DO k=kbcon(i)+1,ktop(i)
                c0=.004
-               if(t(i,k).lt.270.)c0=.002
+!               if(t(i,k).lt.270.)c0=.002
```
remove lines that are changed, do not just comment them out

davegill commented on phys/module_cu_gf_wrfdrv.F:
```diff
@@ -8,7 +8,11 @@ MODULE module_cu_gf_wrfdrv
 use module_cu_gf_deep, only: cup_gf,neg_check,autoconv,aeroevap
 use module_cu_gf_sh, only: cup_gf_sh
-
+!++ GF CTRAN ++ lyy 01/2020
```
go through all of the files and remove these comments

davegill commented on phys/module_cu_gf_wrfdrv.F:
```diff
@@ -218,8 +254,12 @@ SUBROUTINE GFDRV(spp_conv,pattern_spp_conv,field_conv,       &
      real,    dimension (its:ite,kts:kte) ::  qcheck
-
-
+!++ GF CTRAN ++ lyy 01/2020
```
Initially there were a couple of blank lines to separate the declarations from the executable code. Do not change change code for no reason.

davegill commented on phys/module_cumulus_driver.F:
several more places in this file where blank lines have been removed

davegill commented on phys/module_cumulus_driver.F:
```diff
@@ -1317,6 +1344,17 @@ SUBROUTINE cumulus_driver(grid                                     &
                ,do_capsuppress=do_capsuppress                   &
                ,cap_suppress_loc=cap_suppress_loc               &
 #endif 
+!++ GF CTRAN ++ lyy 01/2020
```
go ahead and follow the spacing of the "&" character in the surrounding code

liyunyao replied:
Hi Dave, I have finished making changes according to your comments. But I find that I cannot modify this pr now. Can I just create another pr to release-4.2?

davegill commented:
@liyunyao 
> If you remove module_HLaw.F from the chem dir, it should be now in the share dir (based on your makefile mods). Why is that file not listed?

I now see where those are listed in the commit message, my mistake.

liyunyao commented:
Hi Dave, I have finished making changes according to your comments. But I find that I cannot modify this pr now. Can I just create another pr to release-4.2?

davegill commented:
@liyunyao 
We would prefer if you would keep using this PR. We have quite a few comments in this that would be inconvenient to reproduce.

On your local machine
```
git clone https://github.com/liyunyao/WRF WRF_again
cd WRF_again
git checkout develop
-- make changes --
git add the_files_that_you_changed
git commit
git push -u origin develop
```

liyunyao commented:
> @liyunyao
> We would prefer if you would keep using this PR. We have quite a few comments in this that would be inconvenient to reproduce.
> 
> On your local machine
> 
> ```
> git clone https://github.com/liyunyao/WRF WRF_again
> cd WRF_again
> git checkout develop
> -- make changes --
> git add the_files_that_you_changed
> git commit
> git push -u origin develop
> ```

Hi Dave, 

I just git clone the latest WRF. I saw all the changes are already in the WRF. Do I need to make more changes?

Thanks,
Yunyao

davegill commented:
@liyunyao 
Yunyao,
There are a number of requests that are on this PR page:
https://github.com/wrf-model/WRF/pull/1121

Have those all been addressed? Those should be included in this PR, via another commit.

liyunyao commented:
> @liyunyao
> Yunyao,
> There are a number of requests that are on this PR page:
> #1121
> 
> Have those all been addressed? Those should be included in this PR, via another commit.

I tried what you suggested. But it shows the following error:
error: failed to push some refs to 'https://github.com/liyunyao/WRF'

jordanschnell commented:
Hi @liyunyao, do you have any update on the requested changes?

liyunyao commented:
Hi,

I cannot change the code anymore. But all the request are only related to the comment lines. Can we just use this version. If you want to delete these comment lines, I can submit another PR. 

Thanks,
Yunyao

> On Oct 28, 2020, at 4:18 PM, jordanschnell <notifications@github.com> wrote:
> 
> 
> Hi @liyunyao <https://github.com/liyunyao>, do you have any update on the requested changes?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/wrf-model/WRF/pull/1121#issuecomment-718184276>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AOSLG77CVHPRTW5NTZ6SHKTSNB4BPANCNFSM4LCSCJFQ>.
>

davegill commented:
@liyunyao 
1. Many of the regression tests are failing. Please locally build your branch of the source code for chemistry, no chemistry, and for NMM. Once those tests are successful, please reply in one of these comments. Until then, we are turning off automated testing for this PR.

jordanschnell commented:
@liyunyao @davegill This is my fault, I was removing the commented lines and must have zapped something that wasn't supposed to be. It had passed all of the checks before. Is there a way to revert my commits so I can re-remove those comments before reg testing?

jordanschnell commented:
@liyunyao As you see, there are some problems with the regression tests. Were you able to build this branch with WRF_CHEM=0?

liyunyao commented:
Hi Jordan,

It can compiled without WRF-Chem. I make some changes to it. Do you use the new version of my code. The old version cannot. The new version is okay. I can send you the new code again. I think I deleted the code on Github.


> On Dec 17, 2020, at 10:12 AM, jordanschnell <notifications@github.com> wrote:
> 
> 
> @liyunyao <https://github.com/liyunyao> As you see, there are some problems with the regression tests. Were you able to build this branch with WRF_CHEM=0?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/wrf-model/WRF/pull/1121#issuecomment-747497996>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AOSLG77FFZD4HNNIUVMKGKLSVINU7ANCNFSM4LCSCJFQ>.
>

jordanschnell commented:
Hi @liyunyao,

If the code is not the what is here in this PR, then I do not have it. Can you commit the changes here? Were there substantial changes? If not, I can make them directly here. Otherwise, please send me the code or list of changes. Thanks.

davegill commented:
@liyunyao @jordanschnell 
If this code is not for consideration, how about we close this PR, and open up a new PR with the new source code?

liyunyao commented:
Hi,

Here are the codes. I am sorry, I am not very familiar with Github, that’s why I deleted the code carelessly. 
https://drive.google.com/drive/folders/1t1GR7TnrAy7udgWt5BTkT80-PejrmLb4?usp=sharing

Thanks,
Yunyao


> On Dec 17, 2020, at 10:24 AM, jordanschnell <notifications@github.com> wrote:
> 
> 
> Hi @liyunyao <https://github.com/liyunyao>,
> 
> If the code is not the what is here in this PR, then I do not have it. Can you commit the changes here? Were there substantial changes? If not, I can make them directly here. Otherwise, please send me the code or list of changes. Thanks.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/wrf-model/WRF/pull/1121#issuecomment-747506579>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AOSLG73T66CQV6WE3YQLYMTSVIPDPANCNFSM4LCSCJFQ>.
>

jordanschnell commented:
Hi @liyunyao 

Please take a look at PR #1354. It is this PR but instead based on the latest WRF master branch. Importantly, an additional file needed modification (dyn_em/module_first_rk_step_part1.F) for the code to pass the regression tests -- I'm also curious how your experiments worked without these changes. Please take a look, and preferably grab the code to test:

git clone https://github.com/wrf-model/WRF.git
git remote add jordanschnell https://github.com/wrf-model/WRF.git
git fetch origin pull/1354/head:pull_1354
git checkout pull_1354

Your promptness is appreciated in order to make sure this gets into the next release. Thanks.

liyunyao commented:
Hi Jordan,

Sorry for the late reply. I am out of office last week.

Yes, sorry I forget to change the code in the dyn_em/module_first_rk_step_part1.F. I saw your changes in the dyn_em/module_first_rk_step_part1.F line 1297-1300. They look good to me.

I have another concern, but it is not related to my modification. In the original dyn_em/module_first_rk_step_part1.F, the chem_opt is set in the cumulus_driver if (wrf_chem == 1) (see line 1295 in dyn_em/module_first_rk_step_part1.F). However, in the original phys/module_cumulus_driver.F, the chem_opt is set if (em_core == 1) (see line 71 in phys/module_cumulus_driver.F).

Thanks,
Yunyao

> On Dec 30, 2020, at 2:47 PM, jordanschnell <notifications@github.com> wrote:
> 
> 
> Hi @liyunyao <https://github.com/liyunyao>
> Please take a look at PR #1354 <https://github.com/wrf-model/WRF/pull/1354>. It is this PR but instead based on the latest WRF master branch. Importantly, an additional file needed modification (dyn_em/module_first_rk_step_part1.F) for the code to pass the regression tests -- I'm also curious how your experiments worked without these changes. Please take a look, and preferably grab the code to test:
> 
> git clone https://github.com/wrf-model/WRF.git <https://github.com/wrf-model/WRF.git>
> git remote add jordanschnell https://github.com/wrf-model/WRF.git <https://github.com/wrf-model/WRF.git>
> git fetch origin pull/1354/head:pull_1354
> git checkout pull_1354
> 
> Your promptness is appreciated in order to make sure this gets into the next release. Thanks.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/wrf-model/WRF/pull/1121#issuecomment-752737296>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AOSLG77HBTCM63JVTVRLJVTSXN7WRANCNFSM4LCSCJFQ>.
>

jordanschnell commented:
> Hi Jordan, Sorry for the late reply. I am out of office last week. Yes, sorry I forget to change the code in the dyn_em/module_first_rk_step_part1.F. I saw your changes in the dyn_em/module_first_rk_step_part1.F line 1297-1300. They look good to me. I have another concern, but it is not related to my modification. In the original dyn_em/module_first_rk_step_part1.F, the chem_opt is set in the cumulus_driver if (wrf_chem == 1) (see line 1295 in dyn_em/module_first_rk_step_part1.F). However, in the original phys/module_cumulus_driver.F, the chem_opt is set if (em_core == 1) (see line 71 in phys/module_cumulus_driver.F). Thanks, Yunyao
> […](#)
> On Dec 30, 2020, at 2:47 PM, jordanschnell ***@***.***> wrote: Hi @liyunyao <https://github.com/liyunyao> Please take a look at PR #1354 <#1354>. It is this PR but instead based on the latest WRF master branch. Importantly, an additional file needed modification (dyn_em/module_first_rk_step_part1.F) for the code to pass the regression tests -- I'm also curious how your experiments worked without these changes. Please take a look, and preferably grab the code to test: git clone https://github.com/wrf-model/WRF.git <https://github.com/wrf-model/WRF.git> git remote add jordanschnell https://github.com/wrf-model/WRF.git <https://github.com/wrf-model/WRF.git> git fetch origin pull/1354/head:pull_1354 git checkout pull_1354 Your promptness is appreciated in order to make sure this gets into the next release. Thanks. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <[#1121 (comment)](https://github.com/wrf-model/WRF/pull/1121#issuecomment-752737296)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AOSLG77HBTCM63JVTVRLJVTSXN7WRANCNFSM4LCSCJFQ>.

Hi @liyunyao,

Thanks for checking. As for your concern, I noticed that too. You can see that in one of my original commits, I tried to fix this, but it was not passing the regression tests so I had to revert it (https://github.com/wrf-model/WRF/pull/1354/commits/821e66c2b233e9278dacca30b5daa5b64acf3d68). At this point I'm going to keep it how it was originally for this PR since there have not been error reports. It should be as simple as switching what pre-compiler flag chem_opt is under. I will continue to look and get it switched if necessary. 

Jordan
